pkg/apis/provisioning/v1alpha5/constraints.go

```diff
@@ -47,26 +49,62 @@ type Provider = runtime.RawExtension
 
 // ValidatePod returns an error if the pod's requirements are not met by the constraints
 func (c *Constraints) ValidatePod(pod *v1.Pod) error {
+	p := pod.DeepCopy()
+	// The soft preference may conflict with the requirements.
+	// Remove soft constraints/ preferences
+	if p.Spec.Affinity != nil && p.Spec.Affinity.NodeAffinity != nil {
+		p.Spec.Affinity.NodeAffinity.PreferredDuringSchedulingIgnoredDuringExecution = nil
+	}
 	// Tolerate Taints
-	if err := c.Taints.Tolerates(pod); err != nil {
+	if err := c.Taints.Tolerates(p); err != nil {
 		return err
 	}
+	requirements := NewPodRequirements(p)
+	// Test if labels are allowed
+	for key := range requirements.Keys() {
+		if err := IsRestrictedLabel(key); err != nil && !PodLabelExceptions.Has(key) {
+			return err
+		}
+	}
 	// Test if pod requirements are valid
-	requirements := NewPodRequirements(pod)
 	if err := requirements.Validate(); err != nil {
 		return fmt.Errorf("invalid requirements, %w", err)
 	}
-	// Test if pod requirements are compatible
+	// Test if pod requirements are compatible to the provisioner
 	if errs := c.Requirements.Compatible(requirements); errs != nil {
 		return fmt.Errorf("incompatible requirements, %w", errs)
 	}
 	return nil
 }
 
 func (c *Constraints) Tighten(pod *v1.Pod) *Constraints {
+	requirements := c.Requirements.Add(NewPodRequirements(pod).Requirements...)
+	labels := map[string]string{}
+	for key, value := range c.Labels {
+		labels[key] = value
+	}
+	for key := range requirements.Keys() {
+		if !IsRestrictedNodeLabel(key) {
+			values := requirements.Get(key)
+			if !values.IsComplement() && !values.IsEmpty() {
```

Comment:
Can we abstract this inside a requirements helper? 
```
requirements.PickAny(key)
```

Reply:
Since we only pick a value for complement set or Exists operator, when we abstract this inside the requirement helper, we still need all these conditions in Constraints.go. It looks to me like unnecessary duplications.